webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/MergeNodeProperties.scala

```diff
@@ -349,87 +314,94 @@ object MergeNodeProperties {
-                       case Some(g) => buildUp(g, all)
-                     }
-                   }
-      } yield {
-        InvertedTree(root, parents)
-      }
+        _ <- parents.traverse { p =>
+               // P -> S
+               try {
+                 Right(graph.addEdge(p, child))
+               } catch {
+                 case ex: Exception => // At least NPE, IllegalArgEx, UnsupportedException...
+                   Left(SystemError(s"Error when trying to build group hierarchy of group '${child.groupName}' " +
+                                    s"(${child.groupId.value}) for parent '${p.groupName}' (${p.groupId.value})", ex))
+               }
+             }
+        g <- parents match {
+               case Nil  => Right(graph)
+               case h::t => recAddEdges(graph, h, t)
+             }
+      } yield g
     }
 
-    searchDown(group, allTargets)
+    // we are in java highly mutable land - this algo can't be parallelized without going to ZIO
+    val graph = new DefaultDirectedGraph[GroupProp, DefaultEdge](classOf[DefaultEdge])
+    val groupList = groups.values.toList
+    for {
+      // we need to proceed in two pass because all vertices need to be added before edges
+      // add veritce
+      _      <- groupList.traverse { group =>
+                  try {
+                    Right(graph.addVertex(group))
+                  } catch {
+                    case ex: Exception => // At least NPE, IllegalArgEx, UnsupportedException...
+                      Left(SystemError(s"Error when trying to build group hierarchy of group '${group.groupName}' ${group.groupId.value})", ex))
+                  }
+                }
+      // add edges
+      _      <- groupList.traverse { group =>
+                  for {
+                    parents <- group.parentGroups.traverse { p =>
+                                 groups.get(p) match {
+                                   case None    =>
+                                     Left(Inconsistency(s"Error when looking for parent group '${p}' of group '${group.groupName}' " +
+                                                        s"[${group.groupId.value}]. Please check criterium for that group."))
+                                   case Some(g) =>
+                                     Right(g)
+                                 }
+                               }
+                               // reverse parent here, because we start from the childiest
+                    _       <- recAddEdges(graph, group, parents.reverse)
+                  } yield ()
+                }
+      // get connected components
+      sets   <- try {
+                  Right(new ConnectivityInspector(graph).connectedSets().asScala)
+                } catch {
+                  case ex: Exception => Left(SystemError(s"Error when looking for groups hierarchies", ex))
+                }
+      sorted <- sets.toList.traverse { set =>
+                  try {
```

Comment:
Could be replaced by a IOResult.effect (errorMessage) { the block of code).either

Reply:
no, same than earlier, I want to avoir zio runtime here